hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -2732,3 +2741,228 @@ func reconcileMachineConfigServerService(svc *corev1.Service) error {
+	}
+
+	// The deployment depends on the kubeconfig being reported.
+	if hcp.Status.KubeConfig != nil {
+		// Resolve the kubeconfig secret for machine-approver
+		kubeconfigSecret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: hcp.Namespace,
+				Name:      "service-network-admin-kubeconfig",
+			},
+		}
+		err = r.Client.Get(ctx, client.ObjectKeyFromObject(kubeconfigSecret), kubeconfigSecret)
+		if err != nil {
+			return fmt.Errorf("failed to get hosted controlplane kubeconfig secret %q: %w", kubeconfigSecret.Name, err)
+		}
+
+		// Reconcile machine-approver deployment
+		image := imageMachineApprover
+		if _, ok := hcluster.Annotations[hyperv1.MachineApproverImage]; ok {
+			image = hcluster.Annotations[hyperv1.MachineApproverImage]
+		}
+		deployment := machineapprover.Deployment(controlPlaneNamespace.Name)
+		_, err = controllerutil.CreateOrUpdate(ctx, r.Client, deployment, func() error {
+			return reconcileMachineApproverDeployment(deployment, hcluster, sa, kubeconfigSecret, config, image, imageMachineApproverVersion)
+		})
+		if err != nil {
+			return fmt.Errorf("failed to reconcile machine-approver deployment: %w", err)
+		}
+	}
+
+	return nil
+}
+
+type ClusterMachineApproverConfig struct {
+	NodeClientCert NodeClientCert `json:"nodeClientCert,omitempty"`
+}
+type NodeClientCert struct {
+	Disabled bool `json:"disabled,omitempty"`
+}
+
+func reconcileMachineApproverConfig(cm *corev1.ConfigMap) error {
+	// disable the client cert csr approval
+	// will only use cluster-machine-approver for kubelet serving cert
+	cfg := ClusterMachineApproverConfig{
+		NodeClientCert: NodeClientCert{
+			Disabled: true,
+		},
+	}
+	cfgJSON, err := json.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+	b, err := yaml.JSONToYAML(cfgJSON)
+	if err != nil {
+		return err
+	}
+	if cm.Data == nil {
+		cm.Data = make(map[string]string)
+	}
+	cm.Data["config.yaml"] = string(b)
+
+	return nil
+}
+
+func reconcileMachineApproverRole(role *rbacv1.Role) error {
+	role.Rules = []rbacv1.PolicyRule{
+		{
+			APIGroups: []string{"cluster.x-k8s.io"},
+			Resources: []string{"machines", "machines/status"},
+			Verbs:     []string{"get", "list", "watch"},
+		},
+	}
+	return nil
+}
+
+func reconcileMachineApproverRoleBinding(binding *rbacv1.RoleBinding, role *rbacv1.Role, sa *corev1.ServiceAccount) error {
+	binding.RoleRef = rbacv1.RoleRef{
+		APIGroup: "rbac.authorization.k8s.io",
+		Kind:     "Role",
+		Name:     role.Name,
+	}
+
+	binding.Subjects = []rbacv1.Subject{
+		{
+			Kind:      "ServiceAccount",
+			Name:      sa.Name,
+			Namespace: sa.Namespace,
+		},
+	}
+	return nil
+}
+
+func reconcileMachineApproverDeployment(deployment *appsv1.Deployment, hc *hyperv1.HostedCluster, sa *corev1.ServiceAccount, kubeConfigSecret *corev1.Secret, config *corev1.ConfigMap, machineApproverImage, machineApproverVersion string) error {
+	args := []string{
+		"--config=/var/run/configmaps/config/config.yaml",
+		"-v=4",
```

Comment:
we usually use v=3.

Can we enable leader election here?

Reply:
Didn't find any leader election flags in current machine-approver 😢

Reply:
added a TODO comment for leader election